### ISSUE TITLE
Array constructors fail

### ISSUE THREAD

**Issue by zaikunzhang** [flang-compiler/flang#1227]
Flang fails to compile the following [piece of code](https://github.com/equipez/test_compiler/blob/master/test_flang.f90), which seems standard-conforming. 

```
! test_flang.f90
program test_flang
implicit none

integer, parameter :: m = 2
integer, parameter :: n = 100
integer, parameter :: k = 50
real :: x(m, n)

write (*, *) [x(:, k + 1:n), x(:, 1:k)]

end program test_flang
```

This affects all the flang-based compilers, including the classic flang (7.0.1), Huawei Bisheng flang (Bisheng Compiler 1.3.3.b023), AOCC flang (AMD clang version 13.0.0), and nvfortran (22.2-0). The error message looks like the following.

```
nvfortran --version && nvfortran test_flang.f90 && ./a.out

nvfortran 22.2-0 64-bit target on x86-64 Linux -tp haswell 
NVIDIA Compilers and Tools
Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
NVFORTRAN-S-0152-All dimension specifiers must be ':' (test_flang.f90: 9)
NVFORTRAN-S-0152-All dimension specifiers must be ':' (test_flang.f90: 9)
  0 inform,   0 warnings,   2 severes, 0 fatal for test_flang
```

I am not sure whether this is related to [issue #1200](https://github.com/flang-compiler/flang/issues/1200#issue-1084038205).

**Comment by zhaochuanfeng** [comment]
@zaikunzhang Hi, sorry for so late reply, I reproduced this bug and find out this one is not related to #1212.  We're going to investigate and address this issue.

**Comment by zaikunzhang** [comment]
Hello @zhaochuanfeng , thank you and other colleagues very much for working on this! I look forward to this bug being fixed.

**Comment by kiranchandramohan** [comment]
I think this is fixed by https://github.com/flang-compiler/flang/pull/1232. Could you check @zaikunzhang?

**Comment by zaikunzhang** [comment]
Hello @kiranchandramohan and all! 

Thank you very much for the enormous efforts to fix this issue and #1200 . 

Unfortunately, I cannot make the test on my side, because I have never succeeded in building the Classic Flang on my machine. All my tests were made based on the [binary release that dates back to March 29, 2019](https://github.com/flang-compiler/flang/releases/tag/flang_20190329). I look forward to an updated binary release of the compiler, and I have opened a [new issue](https://github.com/flang-compiler/flang/issues/1247) for this. Thank you for your consideration. 

Anyway, I will close #1200 and #1227 for now. Many thanks again!

Best regards,
Zaikun
